### ISSUE TITLE
Support iterators

### ISSUE THREAD

**Issue by mathiasbynens** [mathiasbynens/Array.from#4]
Since these polyfills were initially meant for ES5 and ES3 environments I didn’t implement support for ES6 iterators – any engines that support iterators probably support `Array.from` as well, right?

Well, just in case there is one that won’t, it would be nice to support iterators anyway, if only for this use case:

```js
Array.from('foo\uD834\uDF06bar') // assuming ES6 StringIterator support
// [ 'f', 'o', 'o', '\uD834\uDF06', 'b', 'a', 'r' ]
```

**Comment by Gozala** [comment]
One use use case is http://nwjs.io it already has ES6 Set and Map but it does not have Array.from

**Comment by shaunc** [comment]
+1 -- I came here because wanted a polyfill that supported Array.from(set) in nodejs 0.12. Any chance of that here?

**Comment by ljharb** [comment]
The `Array.from` implementation in https://github.com/paulmillr/es6-shim supports it just fine, but I'd love to see this one support it as well.

**Comment by wheresrhys** [comment]
> until there exists an implementation with iterators, and without Array.from, we probably shouldn't try to shim that here.

node 0.11.16 --harmony provides your use case

**Comment by ljharb** [comment]
The `--harmony` flag is for unstable, likely broken features, and shouldn't ever be used in a production environment - while we could support it, it's only necessary to support unflagged, stock environments.

**Comment by xsilen** [comment]
I notice about Array.from(Iterator/MapIterator) in Map.values()
In firefox M.values() is Iterator
In chrome M.values() is MapIterator

And the problem in using angular2 in es5 is that angular2 use Array.from( someMapInstance.values() ) to get the array result
so i need to get the result like this
```javascript
var from = function(arrayLike) {
    var C = this;
    if (arrayLike == null) {
        throw new TypeError('`Array.from` requires an array-like object, not `null` or `undefined`');
    }
    /**********************start hack code*********************/
    if(typeof arrayLike === 'object' && (arrayLike.toString() === '[object Map Iterator]' || arrayLike.toString() === '[object Iterator]'))
    {
        var A = new Array();
        var item;
        while( (item = arrayLike.next()).done === false ){
            A.push(item.value);
        }
        return A;
    }
    /**********************end hack code***********************/

    var items = Object(arrayLike);
    var mapping = arguments.length > 1;

    var mapFn, T;
    if (arguments.length > 1) {
        mapFn = arguments[1];
        if (!isCallable(mapFn)) {
            throw new TypeError('When provided, the second argument to `Array.from` must be a function');
        }
        if (arguments.length > 2) {
            T = arguments[2];
        }
    }

    var len = toLength(items.length);
    var A = isCallable(C) ? Object(new C(len)) : new Array(len);
    var k = 0;
    var kValue, mappedValue;
    while (k < len) {
        kValue = items[k];
        if (mapFn) {
            mappedValue = typeof T == 'undefined' ? mapFn(kValue, k) : mapFn.call(T, kValue, k);
        } else {
            mappedValue = kValue;
        }
        defineProperty(A, k, {
            'value': mappedValue,
            'configurable': true,
            'enumerable': true,
            'writable': true
        });
        ++k;
    }
    A.length = len;
    return A;
};
defineProperty(Array, 'from', {
    'value': from,
    'configurable': true,
    'writable': true
});
```
after i modify this, my angular2 DI sample works
```typescript
import {Inject, Injector, bind} from 'angular2/di'

class Engine {
	start(){
		return 'engine start';
	}
}

class Car {
	public engine;
	constructor(@Inject(Engine) engine) {
		this.engine = engine;
	}

	drive(){
		return this.engine.start() + '...' + 'begin drive...';
	}
}

var inj = Injector.resolveAndCreate([
	bind(Car).toClass(Car),
	bind(Engine).toClass(Engine)
]);
var car = inj.get(Car);
console.log(car.drive());
```
there are result in my console, aha
```
engine start...begin drive...
```


so, can you support Iterator?

**Comment by wheresrhys** [comment]
> The --harmony flag is for unstable, likely broken features, and shouldn't ever be used in a production environment

I work at a pretty well respected company and we're starting to use it in production. We're not alone https://github.com/search?q=node+--harmony&ref=searchresults&type=Code&utf8=%E2%9C%93 (angular jumps out at me as a fellow early adopter, but I'm sure if you clicked 'next' a few times you'd find some other familiar names... and remember this is only public repos.)

**Comment by jaspervdg** [comment]
I actually ran into the same problem, trying to get all keys from a Map (node 0.12.2). Also, as an additional use case, if you're using a polyfill for BOTH Map and Array.from, it would be really neat if it all worked the way it should.

**Comment by ljharb** [comment]
@wheresrhys the appeal to authority notwithstanding, it's still a terrible idea. Also, using node 0.11, which doesn't get security updates, is a horrible idea also.

**Comment by amir-arad** [comment]
+1

**Comment by falsandtru** [comment]
How about this?

```
> 'foo\uD834\uDF06bar'.match(/[\uD800-\uDBFF][\uDC00-\uDFFF]|[^\uD800-\uDFFF]/g)
["f", "o", "o", "𝌆", "b", "a", "r"]
```

**Comment by mathiasbynens** [comment]
@falsandtru That would skip lone surrogates.

**Comment by falsandtru** [comment]
Then it would be able to support by `/[\uD800-\uDBFF][\uDC00-\uDFFF]|[\uD800-\uDBFF]|[^\uD800-\uDFFF]/g`.

**Comment by ljharb** [comment]
Closed with #27.

**Comment by ljharb** [comment]
v1.1.0 is now released which should contain all the latest fixes, and will iterate properly in every version of every engine.
